File: Digito Verificador/Alphabetic Check Digit/alphacd.scala
object Main {
  def main(args: Array[String]): Unit = {
    import scala.io.StdIn.readLine

    object TabelaValores {

      def AlphaCD(valor: String): Char = {
        var soma = 0
        var proximo = 0
        val digitos = Array.ofDim[Int](valor.length)

        for (i <- 0 until valor.length) {
          val caractere = valor(i).toUpper
          if (caractere == ' ' || caractere == '.' || caractere == '-') {
            // Ignorar caracteres especiais
          } else if (caractere >= '0' && caractere <= '9') {
            digitos(valor.length - i - 1) = caractere.toInt - '0'.toInt
          } else if (caractere >= 'A' && caractere <= 'Z') {
            digitos(valor.length - i - 1) = (caractere.toInt - 'A'.toInt + 1) % 9
          }

          if ((i + 1) % 2 == 0) {
            digitos(valor.length - i - 1) *= 2
            if (digitos(valor.length - i - 1) >= 10) {
              digitos(valor.length - i - 1) = (digitos(valor.length - i - 1) / 10) + (digitos(valor.length - i - 1) % 10)
            }
          }

          soma += digitos(valor.length - i - 1)
        }

        if (soma % 10 == 0) {
          proximo = soma
        } else {
          proximo = ((soma / 10) + 1) * 10
        }

        (proximo - soma + '0').toChar
      }

      def trimString(str: String): String = {
        var i = 0
        while (i < str.length && str(i) == ' ') {
          i += 1
        }
        if (i < str.length) {
          var j = str.length - 1
          while (str(j) == ' ') {
            j -= 1
          }
          str.substring(i, j + 1)
        } else {
          ""
        }
      }

      def run(): Unit = {
        println("ALPHABETIC CHECK DIGIT")
        println("======================")
        println()

        print("Entre matricula alfanumerica (tamanho livre): ")
        val entrada = readLine()
        println("Dígito verificador para " + entrada + " = " + AlphaCD(entrada))
        println()

        print("Entre matricula e dígito verificador (juntos): ")
        val matricula = readLine()
        val digitoVerificador = matricula.charAt(matricula.length - 1)
        val matriculaSemDigito = trimString(matricula.substring(0, matricula.length - 1))

        if (AlphaCD(matriculaSemDigito) == digitoVerificador) {
          println("Numero valido.")
        } else {
          println("Numero invalido.")
        }
      }
    }

    TabelaValores.run()
  }
}

File: Digito Verificador/Alphabetic Check Digit/alphacd.cs
using System;

namespace AlphabeticCheckDigit
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("ALPHABETIC CHECK DIGIT");
            Console.WriteLine("======================");
            Console.WriteLine();

            Console.Write("Entre matricula alfanumerica (tamanho livre): ");
            string entrada = Console.ReadLine();
            Console.WriteLine("Dígito verificador para " + entrada + " = " + AlphaCD(entrada));
            Console.WriteLine();

            Console.Write("Entre matricula e dígito verificador (juntos): ");
            string matricula = Console.ReadLine();
            char digitoVerificador = matricula[matricula.Length - 1];
            string matriculaSemDigito = matricula.Substring(0, matricula.Length - 1).Trim();

            if (AlphaCD(matriculaSemDigito) == digitoVerificador)
            {
                Console.WriteLine("Numero valido.");
            }
            else
            {
                Console.WriteLine("Numero invalido.");
            }
        }

        public static char AlphaCD(string valor)
        {
            int soma = 0;
            int proximo = 0;
            int[] digitos = new int[valor.Length];

            for (int i = 0; i < valor.Length; i++)
            {
                char caractere = char.ToUpper(valor[i]);
                if (caractere == ' ' || caractere == '.' || caractere == '-')
                {
                    // Ignorar caracteres especiais
                    continue;
                }
                else if (caractere >= '0' && caractere <= '9')
                {
                    digitos[valor.Length - i - 1] = caractere - '0';
                }
                else if (caractere >= 'A' && caractere <= 'Z')
                {
                    digitos[valor.Length - i - 1] = (caractere - 'A' + 1) % 9;
                }

                if ((i + 1) % 2 == 0)
                {
                    digitos[valor.Length - i - 1] *= 2;
                    if (digitos[valor.Length - i - 1] >= 10)
                    {
                        digitos[valor.Length - i - 1] = (digitos[valor.Length - i - 1] / 10) + (digitos[valor.Length - i - 1] % 10);
                    }
                }

                soma += digitos[valor.Length - i - 1];
            }

            if (soma % 10 == 0)
            {
                proximo = soma;
            }
            else
            {
                proximo = ((soma / 10) + 1) * 10;
            }

            return (char)(proximo - soma + '0');
        }
    }
}
